### ISSUE TITLE
Deis Controller API

### ISSUE THREAD

**Issue by nasali** [deis/deis#4073]
Hi there,

So I'm trying to use the Controller API to create an app via a REST call like so:

```
$ curl -v -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: token 557df0153faaaae3a5f17fccwer743b1dbbbbabb' -d '{"id":"myapp"}' -X POST http://deis/v1/apps/
```
Which succeeded in creating the app. Then, I attempted see the configuration variables on it via this REST call:
```
$ curl -v -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: token 557df0153faaaae3a5f17fccwer743b1dbbbbabb' -X GET http://deis/v1/apps/myapp/config/
```
Which worked fabulously well and responded with the following:
```
{"uuid":"09e4bfc2-cb36-4214-9172-11d3e56667c7","app":"myapp","owner":"admin","values":{},"memory":{},"cpu":{},"tags":{},"created":"2015-07-20T21:42:12UTC","updated":"2015-07-20T21:42:12UTC"}
```

However, attempting to set my own configuration variable:
```
$ curl -v -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: token 557df0153faaaae3a5f17fccwer743b1dbbbbabb' -d '{"values":{"FOO":"BAR"}' -X POST http://deis/v1/apps/myapp/config/
```
Fails with this response:
```
< HTTP/1.1 400 Bad Request
* Server nginx/1.9.0 is not blacklisted
< Server: nginx/1.9.0
< Date: Mon, 20 Jul 2015 21:53:01 GMT
< Content-Type: text/html
< Content-Length: 172
< Connection: close
< 
<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.9.0</center>
</body>
</html>
```

What am I missing?

Thanks in advance!

**Comment by carmstrong** [comment]
At first glance, looks like a bracket mismatch: `-d '{"values":{"FOO":"BAR"}'`

**Comment by nasali** [comment]
Thanks for the prompt response. I appreciate it. 

You are correct in that the bracket I had posted was indeed mismatched, but that's because I was trying to use dummy values which I edited on the fly and missed that closing bracket. But the actual value I'm sending in the payload looks like this: 
```
'{"values":{"BUILDPACK_URL":"https://github.com/nasali/oracle-oci8-heroku-buildpack"}}'
```
Which seems to consistently result in the 400 error up there, while ```'{"values":{"FOO":"BAR"}}'``` does work.

Is the BUILDPACK_URL environment variable special somehow? Or is there some escape magic I'm missing?

Thanks again.

**Comment by nasali** [comment]
So just a quick update, I saw the same response using the CLI like so:
```
$ deis config:set BUILDPACK_URL="https://github.com/nasali/oracle-oci8-buildpack" -a uphill-uptowner
Creating config... 400 Bad Request
<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.9.0</center>
</body>
</html>
```

**Comment by carmstrong** [comment]
@nasali Out of curiosity, what if you omit the quotes from the URL?

**Comment by nasali** [comment]
@carmstrong I had tried single quotes, double quotes. I just tried no quotes in the folder where I cloned the example-ruby-sinatra app mentioned in the Deis documentation:
```
$ deis config:set BUILDPACK_URL=https://github.com/nasali/oracle-oci8-buildpack
Creating config... 400 Bad Request
<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.9.0</center>
</body>
</html>
```

If I take out the "http://" part, then it seems to work. But the custom buildpack is never used when I do a git push:
```
 deis config:set BUILDPACK_URL=github.com/nasali/oracle-oci8-buildpack
Creating config... done, v3

=== unbent-seashore
DEIS_APP: unbent-seashore
BUILDPACK_URL: github.com/nasali/oracle-oci8-buildpack
```

Am I missing something obvious?

**Comment by bacongobbler** [comment]
@nasali is there any logs from `deisctl journal controller` which may explain the odd error here?

**Comment by nasali** [comment]
@bacongobbler So I did a deisctl journal controller and then ran the following:
```
$ url -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: token 557df0153faaaae3a5f17fccwer743b1dbbbbabb' -d '{"values":{"FOO":"BAR"}}' -X POST http://deis/v1/apps/unbent-seashore/config/
{"uuid":"3a43de46-51d4-49b7-9400-bf3509ea1cc8","app":"unbent-seashore","owner":"ap_git","values":{"DEIS_APP":"unbent-seashore","FOO":"BAR","BUILDPACK_URL":"github.com/nasali/oracle-oci8-buildpack"},"memory":{},"cpu":{},"tags":{},"created":"2015-07-21T16:51:30UTC","updated":"2015-07-21T16:51:30UTC"}
```
The journal window showed the following:
```
Jul 21 16:51:30 deis-3 sh[22301]: INFO unbent-seashore: config unbent-seashore-3a43de4 updated
Jul 21 16:51:30 deis-3 sh[22301]: INFO unbent-seashore: release unbent-seashore-v6 created
Jul 21 16:51:56 deis-3 sh[22301]: 10.132.205.67 "POST /v1/apps/unbent-seashore/config/ HTTP/1.0" 201 299 "curl/7.37.1"

```
Then I tried the call again to set the BUILDPACK_URL like this:
```
$ curl -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: token 557df0153faaaae3a5f17fccwer743b1dbbbbabb' -d '{"values":{"BUILDPACK_URL":"https://github.com/nasali/oracle-oci8-buildpack"}}' -X POST http://deis/v1/apps/unbent-seashore/config/

<html>
<head><title>400 Bad Request</title></head>
<body bgcolor="white">
<center><h1>400 Bad Request</h1></center>
<hr><center>nginx/1.9.0</center>
</body>
</html>
```
But I saw nothing logged in the deisctl journal window.

The services seem fine as far as I can tell:
```
deisctl list
UNIT				MACHINE				LOAD	ACTIVE	SUB
deis-builder.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-controller.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-database.service		38b3b601.../10.10.10.34	loaded	active	running
deis-logger.service		38b3b601.../10.10.10.34	loaded	active	running
deis-logspout.service		0996259c.../10.10.10.67	loaded	active	running
deis-logspout.service		38b3b601.../10.10.10.34	loaded	active	running
deis-logspout.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-publisher.service		0996259c.../10.10.10.67	loaded	active	running
deis-publisher.service		38b3b601.../10.10.10.34	loaded	active	running
deis-publisher.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-registry@1.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-router@1.service		38b3b601.../10.10.10.34	loaded	active	running
deis-router@2.service		ca814ba3.../10.10.10.142	loaded	active	running
deis-router@3.service		0996259c.../10.10.10.67	loaded	active	running
deis-store-daemon.service	0996259c.../10.10.10.67	loaded	active	running
deis-store-daemon.service	38b3b601.../10.10.10.34	loaded	active	running
deis-store-daemon.service	ca814ba3.../10.10.10.142	loaded	active	running
deis-store-gateway@1.service	ca814ba3.../10.10.10.142	loaded	active	running
deis-store-metadata.service	0996259c.../10.10.10.67	loaded	active	running
deis-store-metadata.service	38b3b601.../10.10.10.34	loaded	active	running
deis-store-metadata.service	ca814ba3.../10.10.10.142	loaded	active	running
deis-store-monitor.service	0996259c.../10.10.10.67	loaded	active	running
deis-store-monitor.service	38b3b601.../10.10.10.34	loaded	active	running
deis-store-monitor.service	ca814ba3.../10.10.10.142	loaded	active	running
deis-store-volume.service	0996259c.../10.10.10.67	loaded	active	running
deis-store-volume.service	38b3b601.../10.10.10.34	loaded	active	running
deis-store-volume.service	ca814ba3.../10.10.10.142	loaded	active	running

```
Should I try restarting the controller? Or the platform?

Thanks again.

**Comment by bacongobbler** [comment]
aha! The 400 response may be unrelated to the controller side of things. Since nothing is in the logs, this may be a router issue. Which platform are you running this from? AWS?

**Comment by aledbf** [comment]
@nasali is https://github.com/nasali/oracle-oci8-buildpack a private repo?

**Comment by nasali** [comment]
@bacongobbler Okay, so you're thinking the router isn't liking my configuration value? If so, where can I find more info on how to debug the reason the router is tanking on this value? I just restarted the platform twice and tried again with the same 400 error result both times.
And, I'm running a three node cluster on Digital Ocean and not on AWS if that makes any difference.

@aledbf No, the repo is public, but the URL I pasted was missing 'heroku' so the correct value should be https://github.com/nasali/oracle-oci8-heroku-buildpack. Thanks for pointing that out, though.

Many thanks.

**Comment by bacongobbler** [comment]
> And, I'm running a three node cluster on Digital Ocean and not on AWS if that makes any difference.

Yes. The reason why I ask is due to https://botbot.me/freenode/deis/2015-06-02/?msg=40790427&page=3, which came from #3096.

**Comment by nasali** [comment]
@bacongobbler Thank you for the tips. 

I ended up digging through the router journal and finding NAXSI_FMT logs with my IP address listed with a block=1, which seemed to be triggered only when the value contained 'https://' - I find that part odd still as I'd expect all calls to fail if I'm blacklisted not just some. Anyways, I tried this:
```
$ deisctl config router set firewall/enabled=false
```
And now the API call to set the BUILDPACK_URL=https://github.com/nasali/oracle-oci8-heroku-buildpack works as advertised.

Thanks again for your stellar and helpful response.
